mlir/lib/Catalyst/Transforms/TBAAPatterns.cpp

```diff
@@ -22,16 +23,52 @@
 #include "mlir/Dialect/LLVMIR/LLVMDialect.h"
 
 #include "Catalyst/Transforms/TBAAUtils.h"
+#include "llvm/Support/Casting.h"
 
 using namespace mlir;
 
 namespace catalyst {
-void setTag(mlir::Type baseType, catalyst::TBAATree *tree, mlir::MLIRContext *ctx,
+
+bool isFromExtractAlignedPointerAsIndexOp(Operation *op)
+{
+    if (isa<memref::ExtractAlignedPointerAsIndexOp>(op)) {
+        return true;
+    }
+    auto prevOp = op->getPrevNode();
+    if (prevOp) {
+        return isFromExtractAlignedPointerAsIndexOp(prevOp);
+    }
+    return false;
+}
+
+bool isInsideDeallocHelper(Operation *op)
+{
+    auto parentOp = dyn_cast<func::FuncOp>(op->getParentOp());
+    if (parentOp) {
+        auto str = parentOp.getName();
+        llvm::StringRef dellocRef = "dealloc_helper";
+        return str.compare(dellocRef) == 0;
+    }
+
+    llvm::outs() << parentOp.getName() << "\n";
+    return false;
+}
+
+void setTag(mlir::Type baseType, catalyst::TBAATree *tree, Operation *currentOp,
             mlir::LLVM::AliasAnalysisOpInterface newOp)
 {
+    mlir::MLIRContext *ctx = currentOp->getContext();
     mlir::LLVM::TBAATagAttr tag;
     if (isa<IndexType>(baseType) || isa<IntegerType>(baseType)) {
-        tag = tree->getTag("int");
+        // Index can be used as a pointer.
+        if (isa<IndexType>(baseType) &&
+            (isFromExtractAlignedPointerAsIndexOp(currentOp) || isInsideDeallocHelper(currentOp))) {
+            tag = tree->getTag("any pointer");
+        }
+        else {
+            isInsideDeallocHelper(currentOp);
```

Comment:
The deallocation pass inserts `delloc_helper` function and it does not come with any `ExtractAlignedPointerAsIndexOp`. 

The index pointers (like `%63` and `%56` at the below) are fed to dealloc_helper. So the store and load inside `dealloc_helper` should be added with the pointer tag. (The tags with `%63` and `%56` are inserted determined by `ExtractAlignedPointerAsIndexOp`.) Since the deallcation pass produces all these `AlignedPointer`, I think it is safe to make the entire `delloc_helper` use the pointer tag.


```llvm
  %63 = getelementptr inbounds i1, ptr %55, i32 2
  store i1 true, ptr %63, align 1, !tbaa !6
  %64 = ptrtoint ptr %43 to i64
  store i64 %64, ptr %56, align 4, !tbaa !4
  %65 = call ptr @_mlir_memref_to_llvm_alloc(i64 ptrtoint (ptr getelementptr (i1, ptr null, i32 3) to i64))
  %66 = call ptr @_mlir_memref_to_llvm_alloc(i64 ptrtoint (ptr getelementptr (i1, ptr null, i32 1) to i64))
  call void @dealloc_helper(ptr %54, ptr %54, i64 0, i64 3, i64 1, ptr %56, ptr %56, i64 0, i64 1, i64 1, ptr %55, ptr %55, i64 0, i64 3, i64 1, ptr %65, ptr %65, i64 0, i64 3, i64 1, ptr %66, ptr %66, i64 0, i64 1, i64 1)
  %67 = load i1, ptr %65, align 1, !tbaa !6
  br i1 %67, label %68, label %69
```

Reply:
The deallocation pass inserts `delloc_helper` function and it does not come with any `ExtractAlignedPointerAsIndexOp`. 

The index pointers (like `%63` at the below) are fed to dealloc_helper. So the store and load inside `dealloc_helper` should be added with the pointer tag. (The tag with `%63` is inserted determined by `ExtractAlignedPointerAsIndexOp`.) Since the deallcation pass produces all these `AlignedPointer`, I think it is safe to make the entire `delloc_helper` use the pointer tag.


```llvm
  %63 = getelementptr inbounds i1, ptr %55, i32 2
  store i1 true, ptr %63, align 1, !tbaa !6
  %64 = ptrtoint ptr %43 to i64
  store i64 %64, ptr %56, align 4, !tbaa !4
  %65 = call ptr @_mlir_memref_to_llvm_alloc(i64 ptrtoint (ptr getelementptr (i1, ptr null, i32 3) to i64))
  %66 = call ptr @_mlir_memref_to_llvm_alloc(i64 ptrtoint (ptr getelementptr (i1, ptr null, i32 1) to i64))
  call void @dealloc_helper(ptr %54, ptr %54, i64 0, i64 3, i64 1, ptr %56, ptr %56, i64 0, i64 1, i64 1, ptr %55, ptr %55, i64 0, i64 3, i64 1, ptr %65, ptr %65, i64 0, i64 3, i64 1, ptr %66, ptr %66, i64 0, i64 1, i64 1)
  %67 = load i1, ptr %65, align 1, !tbaa !6
  br i1 %67, label %68, label %69
```

Reply:
The deallocation pass inserts `delloc_helper` function and it does not come with any `ExtractAlignedPointerAsIndexOp`. 

The index pointers (like `%66` at the below) are fed to dealloc_helper (`%21`). So the store and load inside `dealloc_helper` should be added with the pointer tag. 


```llvm
  %66 = call ptr @_mlir_memref_to_llvm_alloc(i64 ptrtoint (ptr getelementptr (i1, ptr null, i32 1) to i64))
  call void @dealloc_helper(ptr %54, ptr %54, i64 0, i64 3, i64 1, ptr %56, ptr %56, i64 0, i64 1, i64 1, ptr %55, ptr %55, i64 0, i64 3, i64 1, ptr %65, ptr %65, i64 0, i64 3, i64 1, ptr %66, ptr %66, i64 0, i64 1, i64 1)
  ...

  define void @dealloc_helper(ptr %0, ptr %1, i64 %2, i64 %3, i64 %4, ptr %5, ptr %6, i64 %7, i64 %8, i64 %9, ptr %10, ptr %11, i64 %12, i64 %13, i64 %14, ptr %15, ptr %16, i64 %17, i64 %18, i64 %19, ptr %20, ptr %21, i64 %22, i64 %23, i64 %24) {
  ...
  45:                                               ; preds = %42
  %46 = getelementptr inbounds i1, ptr %21, i64 %43
  store i1 false, ptr %46, align 1, !tbaa !6
  %47 = add i64 %43, 1
  br label %42
  ...
}
```

Reply:
I think this `dealloc_helper` handles index pointers, so we should insert pointer tags inside it.

Reply:
I see, so actually we would need to follow all uses of the buffers/memrefs where the index-typed pointers are stored into. Tricky indeed.